configuration/outbound.md

```diff
@@ -0,0 +1,201 @@
+ - JSON paths that may have undefined properties need to be dealt with differently depending on if you are using a `path` or an `expr`. Given `doc.foo.bar.smang` as a path where any of those properties may not exist in the doc:
+  - If you're using `path` just use the path as is, if any part of the path is `undefined` the resulting value will safely be `undefined`
+  - However, in `expr` you **do** need to handle this situation: `doc.foo && doc.foo.bar && doc.foo.bar.smang`
+- If any of your `expr` expressions throw an exception (for example because you didn't handle potentially `undefined` properties as noted above) your push will fail
+- If any of your `path` declarations result in an `undefined` value and you have not also declared that property optional your push will fail
+
+## Credentials
+
+To securely store credentials, we'll be using CouchDB's config storage, as this is a convenient place that only CouchDB admins can access.
+
+Passwords are stored under the `medic-credentials` section, under the key declared in config.
+
+In the following example:
+
+```json
+{
+  "destination": {
+    "base_url": "https://example.com",
+    "auth": {
+      "type": "basic",
+      "username": "admin",
+      "password_key": "example.com"
+    },
+    "path": "/api/v1/referral"
+  }
+}
+```
+
+We have our key configured to `example.com`. This means that in CouchDB's admin config we would expect to find a password at `medic-credentials/example.com`.
+
+To add the credential to the admin config you need to either [PUT the value using curl](https://docs.couchdb.org/en/stable/api/server/configuration.html#put--_node-node-name-_config-section-key) or similar:
+
+```sh
+curl -x PUT http://admin:pass@localhost:5984/_node/couchdb@127.0.0.1/_config/medic-credentials/example.com -d "the-real-password"
+```
+
+_(Note that `couchdb@127.0.0.1` is the local node name, and may be different for you depending on your setup.)_
+
+You can also add it via Fauxton:
+ - Navigate to [the Config screen](http://localhost:5984/_utils/#/_config)
+ - Click `Add Option`
+ - The `Section` should be `medic-credentials`, the `Name` should be (in this example) `example.com` and the value should be the password
+ - Click `Create`
+ - You should then be able to see your credential in the list of configuration shown\
+
+## Flow
+
+Outbound pushes happen in two stages:
+ - Sentinel picks up the report and runs transitions over it. Any outbound configuration that is relevant (via executing the `relevant_to` expression) will be added to a task queue
+ - Every 5 minutes sentinel will check its task queue. For each outbound push that is queued, sentinel will perform the mapping and attempt to send the resulting payload (via POST) to the configured web address
+ - If the push succeeds it will be taken out of the task queue
+ - If the push fails (i.e. the response is not 2xx) it remains in the task queue, to be tried again in 5 minutes
```

Comment:
Will it ever be taken out of the queue?

Reply:
Not currently no, it will fail forever until it either succeeds or the task record is manually deleted.

If we find that we are getting lots of failures that we *don't* want to continually retry, or we find we'd rather use some kind of backoff or whatever, and we find these things out *before* we move to some more complete system for this kind of thing we should revisit.